### PR TITLE
Providing access only to designated attendee(Security)

### LINKED ISSUE #5608: Ticket PDF: Security Issue, can be accessed by anyone
The ticket PDF for events can be accessed by anyone who can find out the link, e.g. see https://api.eventyay.com/static/media/attendees/tickets/pdf/6ae42516-0a03-4436-b2a6-8a60f6a2b16d/QkJ0VHpPUX/ae067598-2896-4579-b4d9-8d5fa8a34e01.pdf

Expected should only be possible to open it from the actual account.

### PR DESCRIPTION
**Pull request by mrsaicharan1**
Fixes #5608 

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

- [ ] All the functions created/modified in this PR contain relevant docstrings.

#### Short description of what this resolves:
- Only allows the authorized user to access their ticket.

#### Changes proposed in this pull request:

- Checks if the current_user is anonymous and provides access to the ticket if False.

### CODE REVIEW

File: app/templates/pdf/attendees_pdf.html [hunk 1]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (iamareebjamal) [inline app/templates/pdf/attendees_pdf.html]: Unneeded

**Comment by iamareebjamal** [comment]
Better add security at route level and through ACCESS DENIED

**Comment by mrsaicharan1** [comment]
@iamareebjamal I think that would be a better idea. On it.

**Comment by mrsaicharan1** [comment]
In addition to my previous point, I think the main problem was accessing the ticket illegally through the link. I think route level security is already a part of the app as authentication is required to generate or download the tickets. Page level security will be the only thing leftover where the user's session variables and auth status will be checked so that unauthorized individuals won't be able to access it. Your thoughts on this, @iamareebjamal ?

**Comment by iamareebjamal** [comment]
You're confusing authentication with authorization.

**Comment by mrsaicharan1** [comment]
Agreed. I'll implement the route level security.

**Comment by mrsaicharan1** [comment]
At the route level, I haven't been able to find where I could check for the user permissions. At the API level, where can I code the logic? @iamareebjamal Would be great if you could help me out here.

**Comment by iamareebjamal** [comment]
The ticket PDF is allowed to be accessed by static URL. The whole mechanism needs to change

**Comment by iamareebjamal** [comment]
Create a new endpoint to show the PDF and move the PDFs from static folder to some internal folder not accessible to the public

**Comment by mrsaicharan1** [comment]
Alright, will be implementing a new route which only gives access to the
ticket buyer and throws a 403 otherwise.

**Comment by mrsaicharan1** [comment]
I've added a custom decorator which checks for anonymity but it seems that even though the user has been logged in, it still throws an anonymous error. What might be happening here?

**Comment by mrsaicharan1** [comment]
`<flask_login.mixins.AnonymousUserMixin object at 0x7fc7f67c1048>` Gives me an anonymous user object when I print the `current_user` post login

**Comment by mrsaicharan1** [comment]
@iamareebjamal  `<flask_login.mixins.AnonymousUserMixin object at 0x7fc7f67c1048>` Gives me an anonymous user object when I print the `current_user` post login. What might be the cause?

**Comment by mrsaicharan1** [comment]
@CosmicCoder96 @iamareebjamal Please review.

**Comment by iamareebjamal** [comment]
If it's ready for review, please remove wip from the title.

**Comment by iamareebjamal** [comment]
Remove unnecessary changes and resolve hound issues

**Comment by mrsaicharan1** [comment]
@iamareebjamal Done.

**Comment by iamareebjamal** [comment]
See my comment on directory traversal

File: app/__init__.py [hunk 2]
```diff
@@ -40,7 +40,6 @@
 from app.views.celery_ import celery
 from app.templates.flask_ext.jinja.filters import init_filters
 
-
```
Comment (iamareebjamal) [inline app/__init__.py]: Unnecessary change

File: app/__init__.py [hunk 3]
```diff
@@ -250,6 +250,5 @@ def internal_server_error(error):
     return make_response(json.dumps(jsonapi_errors([exc.to_dict()])), exc.status,
                          {'Content-Type': 'application/vnd.api+json'})
 
-
```
Comment (iamareebjamal) [inline app/__init__.py]: Unnecessary change

File: app/__init__.py [hunk 4]
```diff
@@ -140,7 +140,7 @@ def create_app():
         app.register_blueprint(attendee_misc_routes)
         app.register_blueprint(order_misc_routes)
         app.register_blueprint(role_invites_misc_routes)
-
```
Comment (iamareebjamal) [inline app/__init__.py]: Add a line

File: app/api/helpers/order.py [hunk 5]
```diff
@@ -55,7 +55,6 @@ def create_pdf_tickets_for_holder(order):
                               UPLOAD_PATHS['pdf']['ticket_attendee'],
                               dir_path='/static/uploads/pdf/tickets/')
         order.tickets_pdf_url = pdf
-
```
Comment (iamareebjamal) [inline app/api/helpers/order.py]: Unnecessary

File: app/api/helpers/order.py [hunk 6]
```diff
@@ -120,3 +119,4 @@ def create_onsite_attendees_for_order(data):
 
     # delete from the data.
     del data['on_site_tickets']
+
```
Comment (iamareebjamal) [inline app/api/helpers/order.py]: Unnecessary

File: app/api/orders.py [hunk 7]
```diff
@@ -10,7 +10,7 @@
 from app.api.bootstrap import api
 from app.api.data_layers.ChargesLayer import ChargesLayer
 from app.api.helpers.db import save_to_db, safe_query, safe_query_without_soft_deleted_entries
-from app.api.helpers.errors import BadRequestError
+from app.api.helpers.errors import BadRequestError, ForbiddenError
```
Comment (iamareebjamal) [inline app/api/orders.py]: Unnecessary as already reported by hound and you still resolved it after I explicitly unresolved it

File: app/api/orders.py [hunk 8]
```diff
@@ -34,7 +34,6 @@
 
 order_misc_routes = Blueprint('order_misc', __name__, url_prefix='/v1')
 
-
```
Comment (iamareebjamal) [inline app/api/orders.py]: Unnecessary

File: app/templates/pdf/ticket_attendee.html [hunk 9]
```diff
@@ -7,12 +7,10 @@
             size: a4 portrait;
             background-image: url('{{ base_dir }}/static/data/ticket-trans-notext.png');
             margin: 0;
-
```
Comment (iamareebjamal) [inline app/templates/pdf/ticket_attendee.html]: Unnecessary

File: app/templates/pdf/ticket_attendee.html [hunk 10]
```diff
@@ -7,12 +7,10 @@
...
             @frame col1_frame {             /* Content frame 1 */
                 left: 80pt;  top: 30pt;
                 height: 250pt; width: 300pt;
             }
-
```
Comment (iamareebjamal) [inline app/templates/pdf/ticket_attendee.html]: Unnecessary

File: app/templates/pdf/ticket_attendee.html [hunk 11]
```diff
@@ -22,21 +20,16 @@
                 width: 100pt; height: 30pt;
                 -pdf-frame-content: number_content;
             }
-
```
Comment (iamareebjamal) [inline app/templates/pdf/ticket_attendee.html]: Unnecessary

File: app/templates/pdf/ticket_attendee.html [hunk 12]
```diff
@@ -22,21 +20,16 @@
...
-
```
Comment (iamareebjamal) [inline app/templates/pdf/ticket_attendee.html]: Unnecessary

File: app/templates/pdf/ticket_attendee.html [hunk 13]
```diff
@@ -22,21 +20,16 @@
...
              @frame col2_frame {
                 left: 440pt; top: 170pt;
             }
         }
-
```
Comment (iamareebjamal) [inline app/templates/pdf/ticket_attendee.html]: Unnecessary

File: app/templates/pdf/ticket_attendee.html [hunk 14]
```diff
@@ -22,21 +20,16 @@
...
         #qrcode {
             -pdf-frame-break: true;
         }
-
```
Comment (iamareebjamal) [inline app/templates/pdf/ticket_attendee.html]: Unnecessary

File: app/templates/pdf/ticket_attendee.html [hunk 15]
```diff
@@ -22,21 +20,16 @@
...
         .label {
             color: #858585;
         }
-
```
Comment (iamareebjamal) [inline app/templates/pdf/ticket_attendee.html]: Unnecessary

**Comment by iamareebjamal** [comment]
> Done and removed unnecessary changes.

No you didn't

File: app/api/orders.py (hunk 7 above)
Comment (mrsaicharan1) [reply app/api/orders.py]: Did not mean to do it. Will correct all the errors. These changes got reverted somehow. On it.

**Comment by mrsaicharan1** [comment]
@iamareebjamal Fixed the issues. Is there anything else required for this PR?

File: app/api/auth.py [hunk 16]
```diff
@@ -269,3 +272,20 @@ def change_password():
         "name": user.fullname if user.fullname else None,
         "password-changed": True
     })
+
+
+def authorized_access_ticket(ticket_function):
+    @wraps(ticket_function)
+    def wrapper(*args, **kwargs):
+        if current_identity.is_anonymous:
```
Comment (iamareebjamal) [inline app/api/auth.py]: This is not checking if the ticket is bought by the person. Only if the person is logged in?

**Comment by mrsaicharan1** [comment]
I was trying to apply the check if current_user.id ==
Order.query.filter_by(identifier=identifier).first().id but it comes out to
be False. Will try it again though

**Comment by mrsaicharan1** [comment]
I was trying to apply the check if current_user.id ==
Order.query.filter_by(identifier=identifier).first().user.id but it comes
out to be False. Will try it again though

File: app/api/auth.py (hunk 16 above)
Comment (mrsaicharan1) [reply app/api/auth.py]: Check for current user's id also checked now.

Comment (iamareebjamal) [reply app/api/auth.py]: I can't see where it is implemented?

Comment (mrsaicharan1) [reply app/api/auth.py]: Forgot to push changes. Updating right now.

Comment (mrsaicharan1) [reply app/api/auth.py]: Updated.

File: app/api/auth.py [hunk 17]
```diff
@@ -269,3 +273,20 @@ def change_password():
         "name": user.fullname if user.fullname else None,
         "password-changed": True
     })
+
+
+def authorized_access_ticket(ticket_function):
+    @wraps(ticket_function)
+    def wrapper(*args, **kwargs):
+        if current_identity.is_anonymous and current_user.id == Order.query.filter_by(identifier=kwargs['identifier']).first().user.id:
```
Comment (iamareebjamal) [inline app/api/auth.py]: Wrong logic?

Comment (mrsaicharan1) [reply app/api/auth.py]: how do we check if thhe current user is associated with the ticket?

Comment (iamareebjamal) [reply app/api/auth.py]: Here you'll throw an error, by your own logic, 

1. If there is no logged in user, AND
2. Current user's ID matches with the one who placed the order

Seems something wrong with this logic?

Comment (mrsaicharan1) [reply app/api/auth.py]: Just a quick doubt, Is there a possibility that if there is no logged in user, the current_user can still be present in the session?

Comment (iamareebjamal) [reply app/api/auth.py]: No, that's why this does not make any sense

**Comment by mrsaicharan1** [comment]
Yes. I understood where I went wrong. Need to swap the conditions . Will
make the changes and revert.

**Comment by mrsaicharan1** [comment]
@iamareebjamal Is the condition checking right now?

**Comment by iamareebjamal** [comment]
What if your query returns 0 results. I mean if I enter wrong identifier?

**Comment by mrsaicharan1** [comment]
Will be restarting this in a new PR for clarity
